edflow/eval/pipeline.py

```diff
@@ -0,0 +1,658 @@
+        os.makedirs(self.save_root, exist_ok=True)
+
+        self.label_arrs = None
+
+    def after_step(self, step, last_results):
+        # Attention -> This does not work with nested keys!
+        # Fix it in post :)
+        label_vals = {k: _delget(last_results["step_ops"], k) for k in self.lks}
+
+        if self.label_arrs is None:
+            self.label_arrs = {}
+            for k in self.lks:
+                example = label_vals[k][0]
+                ex_shape = list(np.shape(example))
+                shape = [len(self.data_in)] + ex_shape
+                s = "x".join([str(s) for s in shape])
+                dtype = d = example.dtype
+
+                savepath = os.path.join(
+                    self.save_root, "{}-*-{}-*-{}.npy".format(k, s, d)
+                )
+                memmap = np.memmap(savepath, shape=tuple(shape), mode="w+", dtype=dtype)
+                self.label_arrs[k] = memmap
+
+        idxs = np.array(last_results["step_ops"]["index_"], dtype=int)
+
+        for k in self.lks:
+            # Can the inner loop be made a fancy indexing assign?
+            for i, idx in enumerate(idxs):
+                self.label_arrs[k][idx] = label_vals[k][i]
+
+        path_dicts = save_output(self.save_root, last_results, self.sdks)
+
+        if self.data_frame is None:
+            columns = sorted(path_dicts[list(path_dicts.keys())[0]])
+            if len(columns) == 0:
+                # No load heavy logs written out
+                pass
+            else:
+                self.data_frame = pd.DataFrame(columns=columns)
+
+        if self.data_frame is not None:
+            for idx, path_dict in path_dicts.items():
+                self.data_frame.loc[idx] = path_dict
+
+    def at_exception(self, *args, **kwargs):
+        self.save_csv()
+
+    def after_epoch(self, epoch):
+        """Save csv for reuse and then start the evaluation callbacks"""
+        self.save_csv()
+
+        data_out = EvalDataFolder(self.root)
+
+        for n, cb in zip(self.cb_names, self.cbacks):
+            cb_name = "CB: {}".format(n)
+            cb_name = "{a}\n{c}\n{a}".format(a="=" * len(cb_name), c=cb_name)
+            self.logger.info(cb_name)
+            cb(self.root, self.data_in, data_out, self.meta)
+
+    def save_csv(self):
+        csv_path = os.path.join(self.root, "model_output.csv")
+
+        if self.data_frame is not None:
+            self.data_frame = self.data_frame.sort_index()
+            self.data_frame.to_csv(csv_path, index=False)
+        else:
+            with open(csv_path, "w+") as csv_file:
+                csv_file.write("")
+
+        add_meta_data(csv_path, self.meta)
+
+        this_script = os.path.dirname(__file__)
+        cbs = " ".join(self.cb_names)
+
+        self.logger.info(
+            "All data has been produced. You can now also run all"
+            + " callbacks using the following commands:\n"
+            + "cd {}\n".format(this_script)
```

Comment:
will this point to some cryptic installation where edflow is installed? maybe change to the usage of the `edeval` command